tss-esapi/src/structures/buffers/public.rs

```diff
@@ -0,0 +1,501 @@
+    /// [Rsa][`crate::interface_types::algorithm::PublicAlgorithm::Rsa].
+    pub fn with_rsa_parameters(mut self, rsa_parameters: &PublicRsaParameters) -> Self {
+        self.rsa_parameters = Some(*rsa_parameters);
+        self
+    }
+
+    /// Adds the RSA unique identifier for the [Public] structure
+    /// to the builder.
+    ///
+    /// # Details
+    /// This is required if the public algorithm is set to
+    /// [Rsa][`crate::interface_types::algorithm::PublicAlgorithm::Rsa].
+    ///
+    /// The unique identifier is the public key.
+    pub fn with_rsa_unique_identifier(mut self, rsa_unique_identifier: &PublicKeyRsa) -> Self {
+        self.rsa_unique_identifier = Some(rsa_unique_identifier.clone());
+        self
+    }
+
+    /// Adds the keyed hash parameters for the [Public] structure
+    /// to the builder.
+    ///
+    /// # Details
+    /// This is required if the public algorithm is set to
+    /// [KeyedHash][`crate::interface_types::algorithm::PublicAlgorithm::KeyedHash].
+    pub fn with_keyed_hash_parameters(
+        mut self,
+        keyed_hash_parameters: &PublicKeyedHashParameters,
+    ) -> Self {
+        self.keyed_hash_parameters = Some(*keyed_hash_parameters);
+        self
+    }
+
+    /// Adds the keyed hash unique identifier for the [Public] structure
+    /// to the builder.
+    ///
+    /// # Details
+    /// This is required if the public algorithm is set to
+    /// [KeyedHash][`crate::interface_types::algorithm::PublicAlgorithm::KeyedHash].
+    pub fn with_keyed_hash_unique_identifier(
+        mut self,
+        keyed_hash_unique_identifier: &Digest,
+    ) -> Self {
+        self.keyed_hash_unique_identifier = Some(keyed_hash_unique_identifier.clone());
+        self
+    }
+
+    /// Adds the ECC parameters for the [Public] structure
+    /// to the builder.
+    ///
+    /// # Details
+    /// This is required if the public algorithm is set to
+    /// [Ecc][`crate::interface_types::algorithm::PublicAlgorithm::Ecc].
+    pub const fn with_ecc_parameters(mut self, ecc_parameters: PublicEccParameters) -> Self {
+        self.ecc_parameters = Some(ecc_parameters);
+        self
+    }
+
+    /// Adds the ECC unique identifier for the [Public] structure
+    /// to the builder.
+    ///
+    /// # Details
+    /// This is required if the public algorithm is set to
+    /// [Ecc][`crate::interface_types::algorithm::PublicAlgorithm::Ecc].
+    ///
+    /// The unique identifier is a ecc point.
+    pub fn with_ecc_unique_identifier(mut self, ecc_unique_identifier: EccPoint) -> Self {
+        self.ecc_unique_identifier = Some(ecc_unique_identifier);
+        self
+    }
+
+    /// Adds the symmetric cipher parameters for the [Public] structure
+    /// to the builder.
+    ///
+    /// # Details
+    /// This is required if the public algorithm is set to
+    /// [SymCipher][`crate::interface_types::algorithm::PublicAlgorithm::SymCipher].
+    pub const fn with_symmetric_cipher_parameters(
```

Comment:
I will fix that. We should always try to take arguments that cannot be copied by reference.

Reply:
But there are inconsistencies in there. Most of the parameters can be copied(if one should be able to copy them is another story) so they are fine to take by value. Though it is not possible to copy EccPoint so it should be taken by reference. So I will have to look at this more to fix the stuff.

Reply:
I have now changed so that the types that cannot be copied are taken by reference and the rest by value is this the way we want to go our should they all be taken by reference?